subsys/pm/policy.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pm/pm.h>
+#include <pm/policy.h>
+#include <spinlock.h>
+#include <sys_clock.h>
+#include <sys/__assert.h>
+#include <sys/time_units.h>
+#include <sys/atomic.h>
+
+/** State disallow reference counting */
+static atomic_t state_disallow_cnt[PM_STATE_COUNT];
+/** Callback to notify when state allowed status changes */
+static pm_policy_state_allowed_changed_cb_t allowed_changed_cb;
+
+/** Lock to synchronize access to the latency request list. */
+static struct k_spinlock latency_lock;
+/** List of maximum latency requests. */
+static sys_slist_t latency_reqs;
+/** Maximum CPU latency in ticks */
+static int32_t max_latency_ticks = K_TICKS_FOREVER;
+/** Callback to notify when maximum latency changes. */
+pm_policy_latency_changed_cb_t latency_changed_cb;
+
+/** @brief Update maximum allowed latency. */
+static void update_max_latency(void)
+{
+	int32_t new_max_latency_ticks = K_TICKS_FOREVER;
+	struct pm_policy_latency_request *req;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&latency_reqs, req, node) {
+		if ((new_max_latency_ticks == K_TICKS_FOREVER) ||
```

Comment:
maybe we could keep the list sorted for better performance?

Reply:
I'll take a look if it's worth, since this operation is `O(N)`, doing a linear search to insert/update is also `O(N)`. Removal would benefit, though.

Reply:
If we keep list sorted (in order of increased latency), then the stricter requirements will be faster to insert/remove. Note, that our interest lays in getting the most strict value.

Reply:
If we keep the list sorted, we still need to do a search when inserting, which I guess the best we can have is `O(log N)` by partitioning. I'd first wait to see performance numbers before going into micro-optimizations.

Reply:
Agree ... How many latencies we expect to have ? Probably not much that justify
the complexity. Also, this is a linked list, which means that even to do a
binary search there is an overhead to find the middle element and you have to
have access to the last element. So this probably only justify for big lists.